### ISSUE TITLE
No 'Access-Control-Allow-Origin' CORS blocked - Node app with Apache proxy

### ISSUE THREAD

**Issue by allanext** [parse-community/parse-server#6674]
### Issue Description

I had no issues on running client & server on localhost but i'm getting the following when deployed for production. The client is delivered with a CDN and the server with an Apache proxy that redirects to the node app.

This is on the browser console inspector:
> Access to XMLHttpRequest at 'https://showcase-s.domain.com/showcase-s_domain_com/classes/m_settings' 
> from origin 'https://showcase.domain.com' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' 
> header is present on the requested resource.

I'm running multiple applications, each one responds to a single api (eg https://showcase-s.domain.com/showcase-s_domain_com)

I understand it's related to express but I'm wondering if anybody went through this configuration or there are things to consider with Parse server.

This is the Apache server proxy configuration:

```
    ServerName s.domain.com
    ServerAlias *-s.domain.com
    DocumentRoot /var/www/domain2-parse/public

    SSLProxyEngine On
    ProxyPreserveHost On
    ProxyRequests Off

    # This is for websocket requests
    ProxyPass /showcase-s_domain_com wss://127.0.0.1:1337/
    ProxyPassReverse /showcase-s_domain_com wss://127.0.0.1:1337/
    ProxyPass /test-s_domain_com wss://127.0.0.1:1337/
    ProxyPassReverse /test-s_domain_com wss://127.0.0.1:1337/

    # not useful here
    # Header set Access-Control-Allow-Origin "*"
    # Header always set Access-Control-Allow-Methods "POST, PUT, GET, DELETE, OPTIONS"
    # Header always set Access-Control-Allow-Headers "Content-Type"
    # Header add Access-Control-Allow-Headers "origin, x-requested-with, content-type"

    #Header Set Access-Control-Allow-Origin "https://test.domain.com"
    #Header Set Access-Control-Allow-Origin "https://showcase.domain.com"


    #<Directory "/var/www/domain2-parse/">
    #   <IfModule mod_headers.c>
    #       Header set Access-Control-Allow-Origin "*"
    #       Header always set Access-Control-Allow-Methods "POST, PUT, GET, DELETE, OPTIONS"
    #       Header add Access-Control-Allow-Headers "origin, x-requested-with, content-type"
    #       Header always set Access-Control-Allow-Headers "Content-Type"
    #  </IfModule>
    #</Directory>   
    
    # This is for normal requests    
    ProxyPass / https://127.0.0.1:1337/
    ProxyPassReverse / https://127.0.0.1:1337/

    ErrorLog /var/log/httpd/parse.domain.com-error.log
    CustomLog /var/log/httpd/parse.domain.com-access.log combined
    SSLEngine On
    SSLCertificateFile    /etc/letsencrypt/live/domain.com/cert.pem
    SSLCACertificateFile  /etc/letsencrypt/live/domain.com/chain.pem
    SSLCertificateKeyFile /etc/letsencrypt/live/domain.com/privkey.pem
```

As you can see below i've tried in multiple ways to set the Access-Control-Allow-Origin in the express middleware callbacks but I still get the error

Parse server index.js :

```
var express = require('express');
var ParseServer = require('parse-server').ParseServer;
var path = require('path');
var helmet = require('helmet');
var cors = require('cors');

const apps = [
{
  {
    NAME: "Showcase",
    URL: 'showcase.domain.com',
    SERVER_URL: 'https://showcase-s.domain.com/showcase-s_domain_com',
    PARSE_KEY: 'xxx',
    JAVASCRIPT_KEY: 'xxx',
    MASTER_KEY: 'xxx'
  },
  {
    NAME: "Test",
    URL: 'test.domain.com',
    SERVER_URL: 'https://test-s.domain.com/test-s_domain_com',
    PARSE_KEY: 'xxx',
    JAVASCRIPT_KEY: 'xxx',
    MASTER_KEY: 'xxx'
  }
];
var app = express();
app.use(helmet());
var allowedOrigins = ['http://localhost:3000',
                      'https://showcase.domain.com', 
                      'https://showcase-s.domain.com',
                      'https://test.domain.com'
                    ];
app.use(cors({
  credentials: true,
  origin: (origin, callback) => {
    // allow requests with no origin 
    // (like mobile apps or curl requests)
    if(!origin) return callback(null, true);
    if(allowedOrigins.indexOf(origin) === -1){
      var msg = 'The CORS policy for this site does not ' +
                'allow access from the specified Origin.';
      return callback(new Error(msg), false);
    }
    return callback(null, true);
  }
}));
app.options('*', cors()); // Enable CORS-Pre-Flight

app.use( (req, res, next) => {

  // Website you wish to allow to connect
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Origin', 'https://showcase.domain.com');

  // Request methods you wish to allow
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE');

  // Request headers you wish to allow
  res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type');

  // Set to true if you need the website to include cookies in the requests sent
  // to the API (e.g. in case you use sessions)
  res.setHeader('Access-Control-Allow-Credentials', true);

  // Pass to next layer of middleware
  next();
});
app.all('*', cors(), (req, res, next) => {
  res.jsonp({msg: 'This is CORS-enabled for all origins!'});
});

app.all('*', (req, res, next) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.setHeader('Access-Control-Allow-Origin', 'https://showcase.domain.com');
  // Request methods you wish to allow
  res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE');
  // Request headers you wish to allow
  res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type');
  // Set to true if you need the website to include cookies in the requests sent
  // to the API (e.g. in case you use sessions)
  res.setHeader('Access-Control-Allow-Credentials', true);
  // Pass to next layer of middleware
  next();
  
  if ('OPTIONS' == req.method) {
    res.send(200);
  }
  else {
    next();
  }
});

var servers = [];
apps.forEach((conf) => {
  var db = conf.URL.substring(0, (conf.URL.indexOf('.')==-1)?conf.URL.length:conf.URL.indexOf('.')) + '_domain_com'
  console.log("App: "+ conf.URL +" Conncting to database: "+ db);

  api = new ParseServer({
    appName: conf.NAME,
    databaseURI: 'mongodb://127.0.0.1:27017/' + db,
    cloud: process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js',
    verbose: false,
    appId: conf.PARSE_KEY,
    clientKey: conf.JAVASCRIPT_KEY,
    masterKey: conf.MASTER_KEY,
    serverURL:  'https://'+ conf.URL +':1337/' + db, 
    publicServerURL: 'https://'+ conf.URL +':1337/' + db,
    /*protectedFields: {
      _User: {
        // "*": ["email", "sin"],
        // "role:moderator": ["sin"],
        // "role:admin": []
      }
    }, */ 
    liveQuery: {
      classNames: [
        "_User",
        "m_apps", 
        'm_audits',  
        'm_settings'
      ]
    },
    verifyUserEmails: true,
    emailVerifyTokenValidityDuration: 12 * 60 * 60,
    preventLoginWithUnverifiedEmail: true,
    emailAdapter: {
      module: 'parse-server-simple-ses-adapter',
      options: {
          fromAddress: 'bot@domain.com',
          apiKey: 'XXX',                         // SES API Key
          apiSecret: 'WWW',  // SES API Secret
          domain: 'domain.com',                  // A verified domain in SES
          amazon: 'https://email.us-east-1.amazonaws.com' // OPTIONAL: Defaults to us-east-1
      }
    }
  });

  var mountPath = '/' + db;
  app.use(mountPath, api);
  servers.push(api);
})

// Serve static assets from the /public folder
app.use('/public', express.static(path.join(__dirname, '/public')));

// Parse Server plays nicely with the rest of your web routes
app.get('/', (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', 'https://showcase.domain.com');
  res.setHeader('Access-Control-Allow-Origin', 'https://test.domain.com');
  res.status(200).send('I dream of being a big website!');
});
app.post('/showcase-s_domain_com/classes/m_settings', (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', 'https://showcase.domain.com');
  res.setHeader('Access-Control-Allow-Origin', 'https://test.domain.com');
  res.status(200).send('I dream of being a big website!');
});
app.get('/showcase-s_domain_com/classes/m_settings', (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', 'https://showcase.domain.com');
  res.setHeader('Access-Control-Allow-Origin', 'https://test.domain.com');
  res.status(200).send('I dream of being a big website!');
});
app.post('/showcase-s_domain_com/classes/m_settings', (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.status(200).send('I dream of being a big website!');
});
app.get('/showcase-s_domain_com/classes/m_settings', (req, res) => {
  res.setHeader('Access-Control-Allow-Origin', '*');
  res.status(200).send('I dream of being a big website!');
});

// Certificates and credentials for HTTPS server
var fs = require('fs')
var privateKey  = fs.readFileSync('/etc/letsencrypt/live/domain.com/privkey.pem', 'utf8') //__dirname + 
var certificate = fs.readFileSync('/etc/letsencrypt/live/domain.com/cert.pem', 'utf8')
var ca = fs.readFileSync('/etc/letsencrypt/live/domain.com/chain.pem', 'utf8')
var credentials = {key: privateKey, cert: certificate, ca: ca}

app.enable('trust proxy')
app.set("trust proxy", 1)

var s_port = 1337;
var server = require('https').createServer(credentials, app);
server.listen(s_port, function() {
    var host = server.address().address
    var port = server.address().port
    console.log('domain server listening at https://%s:%s', host, port)
});

// This will enable the Live Query real-time server
ParseServer.createLiveQueryServer(server);
```

### Environment Setup

Parse server is not the latest as I have another issue.. are they any versions requirements for compatibility between parse-server, parse-js and express?

- **Server**
```
  "name": "parse-server",
  "version": "1.0.0",
  "description": "domain Parse Server",
  "main": "index.js",
  "dependencies": {
    "cors": "^2.8.5",
    "express": "^4.17.1",
    "helmet": "^3.22.0",
    "kerberos": "~0.0.x",
    "parse-server": "3.10.0",
    "parse-server-simple-ses-adapter": "^1.0.6"
  },
  "scripts": {
    "start": "node index.js"
  },
  "engines": {
    "node": ">=4.3"
  }
```

- **Parse Client**
```
    ...
    "parse": "2.12.0",
    ...
```

I've also tried to initiate the different apps like this:

```
apps.forEach((conf) => {
  var db = conf.URL.substring(0, (conf.URL.indexOf('.')==-1)?conf.URL.length:conf.URL.indexOf('.')) + '_domain_com'
  console.log("App: "+ conf.URL +" Conncting to database: "+ db);

  api = new ParseServer({
    appName: conf.NAME,
    databaseURI: 'mongodb://127.0.0.1:27017/' + db,
    cloud: process.env.CLOUD_CODE_MAIN || __dirname + '/cloud/main.js',
    verbose: false,
    appId: conf.PARSE_KEY,
    clientKey: conf.JAVASCRIPT_KEY,
    masterKey: conf.MASTER_KEY,
    serverURL:  'https://'+ conf.URL +':1337/' + db, 
    publicServerURL: 'https://'+ conf.URL +':1337/' + db,
    /*protectedFields: {
      _User: {
        // "*": ["email", "sin"],
        // "role:moderator": ["sin"],
        // "role:admin": []
      }
    }, */ 
    liveQuery: {
      classNames: [
        "_User",
        "m_apps", 
        'm_audits',  
        'm_settings',
      ] // List of classes to support for query subscriptions
    },
    verifyUserEmails: true,
    emailVerifyTokenValidityDuration: 12 * 60 * 60,
    preventLoginWithUnverifiedEmail: true,
    emailAdapter: {
      module: 'parse-server-simple-ses-adapter',
      options: {
          fromAddress: 'bot@domain.com',
          apiKey: 'WWW',                         // SES API Key
          apiSecret: 'YYY',  // SES API Secret
          domain: 'domain.com',                  // A verified domain in SES
          amazon: 'https://email.us-east-1.amazonaws.com' // OPTIONAL: Defaults to us-east-1
      }
    }
  });
  var mountPath = '/' + db;
  app.use(mountPath, api, (req, res) => {
    res.setHeader('Access-Control-Allow-Origin', '*');
    res.setHeader('Access-Control-Allow-Methods', 'GET, POST, OPTIONS, PUT, PATCH, DELETE'); // If needed
    res.setHeader('Access-Control-Allow-Headers', 'X-Requested-With,content-type'); // If needed
    res.setHeader('Access-Control-Allow-Credentials', true); // If needed

    res.send('cors problem fixed:)');
  });
  servers.push(api);
})
```

### Logs/Trace

The response from the server seems correct apart the browser CORS exception:

> index > info: Create new client: 33c28dcf-5c79-4213-96e8-2f3f1780c

Any ideas? Thank you and I apologies in advance for going through all the code!

**Comment by allanext** [comment]
I missed a log from apache; looks like it's an issue related to apache server modules and wss

> [Fri May 08 21:34:42.118337 2020] [proxy:warn] [pid 1247] [client A.B.C.D:63169] AH01144: No protocol handler was valid for the URL /showcase-s_domain_com/classes/m_settings (scheme 'wss'). If you are using a DSO version of mod_proxy, make sure the proxy submodules are included in the configuration using LoadModule., referer: https://showcase.domain.com/login

**Comment by allanext** [comment]
For now i've just resolved by avoiding the Apache proxy and running the Parse server directly on https.

**Comment by allanext** [comment]
The error came up also by running directly the node app so not an Apache proxy issue (despite the log). I ended up resolving with this on index.js:

```
var cors = require('cors');

var app = express();
app.use(helmet());
var allowedOrigins = ['http://localhost:8080',
                      'http://127.0.0.1:8080',
                      'https://showcase.domain.com', 
                      'https://test.domain.com'
                    ];
                    
app.use(cors({
  credentials: true,
  origin: (origin, callback) => {
    // allow requests with no origin 
    // (like mobile apps or curl requests)
    if(!origin) return callback(null, true);
    if(allowedOrigins.indexOf(origin) === -1){
      var msg = 'The CORS policy for this site does not ' +
                'allow access from the specified Origin.';
      return callback(new Error(msg), false);
    }
    return callback(null, true);
  }
}));
app.options('*', cors()); // Enable CORS-Pre-Flight
```
